File: src/main/scala/com/github/users/controllers/converter/RequestConverter.scala
package com.github.users.controllers.converter

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.users.model.neo4j.User
import com.linecorp.armeria.common.AggregatedHttpRequest
import com.linecorp.armeria.server.ServiceRequestContext
import com.linecorp.armeria.server.annotation.RequestConverterFunction
import org.slf4j.LoggerFactory

import java.lang.reflect.ParameterizedType
import scala.reflect.ClassTag
import scala.reflect.runtime.universe.*

abstract class RequestConverter[T](implicit val classTag: ClassTag[T]) extends RequestConverterFunction:

  private val logger = LoggerFactory.getLogger(getClass)

  private val mapper: ObjectMapper = new ObjectMapper
  private val classType: Class[_] = classTag.runtimeClass

  override def convertRequest(ctx: ServiceRequestContext,
                              request: AggregatedHttpRequest,
                              expectedResultType: Class[?],
                              expectedParameterizedResultType: ParameterizedType
                             ): AnyRef =
    logger.info("I am in RequestConverter")
    if expectedResultType eq classType then
      return mapper.readValue(request.contentUtf8, classType)
    RequestConverterFunction.fallthrough

end RequestConverter

class UserRequestConverter extends RequestConverter[User]

File: crud/src/main/java/io/github/abhishekghoshh/crud/controller/converter/RequestConverter.java
package io.github.abhishekghoshh.crud.controller.converter;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.linecorp.armeria.common.AggregatedHttpRequest;
import com.linecorp.armeria.server.ServiceRequestContext;
import com.linecorp.armeria.server.annotation.RequestConverterFunction;

import java.lang.reflect.ParameterizedType;

public abstract class RequestConverter<T> implements RequestConverterFunction {

    private final ObjectMapper mapper = new ObjectMapper();
    private final Class<T> classType;

    @SuppressWarnings("unchecked")
    protected RequestConverter() {
        this.classType = (Class<T>) extractTypeParameter(getClass());
    }

    @Override
    public Object convertRequest(
            ServiceRequestContext ctx,
            AggregatedHttpRequest request,
            Class<?> expectedResultType,
            ParameterizedType expectedParameterizedResultType
    ) throws Exception {
        if (expectedResultType.equals(classType)) {
            return mapper.readValue(request.contentUtf8(), classType);
        }
        return RequestConverterFunction.fallthrough();
    }

    // Helper to get T from RequestConverter<T>
    private static Class<?> extractTypeParameter(Class<?> clazz) {
        while (clazz != null) {
            if (clazz.getGenericSuperclass() instanceof ParameterizedType parameterizedType) {
                if (parameterizedType.getActualTypeArguments()[0] instanceof Class<?> typeArg) {
                    return typeArg;
                }
            }
            clazz = clazz.getSuperclass();
        }
        throw new IllegalStateException("Unable to determine type parameter T.");
    }
}